components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/ResendCodeApiServiceImpl.java

```diff
@@ -23,35 +23,163 @@
 import org.wso2.carbon.identity.recovery.IdentityRecoveryClientException;
 import org.wso2.carbon.identity.recovery.IdentityRecoveryConstants;
 import org.wso2.carbon.identity.recovery.IdentityRecoveryException;
+import org.wso2.carbon.identity.recovery.RecoveryScenarios;
+import org.wso2.carbon.identity.recovery.RecoverySteps;
 import org.wso2.carbon.identity.recovery.bean.NotificationResponseBean;
+import org.wso2.carbon.identity.recovery.confirmation.ResendConfirmationManager;
+import org.wso2.carbon.identity.recovery.model.UserRecoveryData;
 import org.wso2.carbon.identity.recovery.signup.UserSelfRegistrationManager;
 import org.wso2.carbon.identity.user.endpoint.Constants;
 import org.wso2.carbon.identity.user.endpoint.ResendCodeApiService;
+import org.wso2.carbon.identity.user.endpoint.dto.PropertyDTO;
 import org.wso2.carbon.identity.user.endpoint.util.Utils;
 import org.wso2.carbon.identity.user.endpoint.dto.ResendCodeRequestDTO;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
 
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
 import javax.ws.rs.core.Response;
 
 public class ResendCodeApiServiceImpl extends ResendCodeApiService {
     private static final Log LOG = LogFactory.getLog(ResendCodeApiServiceImpl.class);
+    private static final String RECOVERY_SCENARIO_KEY = "RecoveryScenario";
 
     @Override
     public Response resendCodePost(ResendCodeRequestDTO resendCodeRequestDTO) {
 
-        String tenantFromContext = (String) IdentityUtil.threadLocalProperties.get().get(Constants.TENANT_NAME_FROM_CONTEXT);
+        String tenantFromContext = getTenantDomainFromContext();
 
         if(StringUtils.isNotBlank(tenantFromContext)) {
             resendCodeRequestDTO.getUser().setTenantDomain(tenantFromContext);
         }
 
+        NotificationResponseBean notificationResponseBean = null;
+        List<PropertyDTO> recoveryScenarioProperty = resendCodeRequestDTO.getProperties();
+
+        if (recoveryScenarioProperty.isEmpty()) {
+            notificationResponseBean = doResendConfirmationCodeForSelfSignUp(notificationResponseBean,
+                    resendCodeRequestDTO);
+        } else {
+            String recoveryScenario = validateRecoveryScenarioPropertyList(recoveryScenarioProperty);
+            if (recoveryScenario != null) {
+                notificationResponseBean = doResendConfirmationCode(recoveryScenario, notificationResponseBean,
+                        resendCodeRequestDTO);
+            }
+        }
+
+        if (notificationResponseBean != null) {
+            if (StringUtils.isBlank(notificationResponseBean.getKey())) {
+                return Response.status(Response.Status.CREATED).build();
+            }
+            return Response.status(Response.Status.CREATED).entity(notificationResponseBean.getKey()).build();
+        } else {
+            return Response.status(Response.Status.CREATED).build();
+        }
```

Comment:
Can we simplify this logic
```
           if (notificationResponseBean == null || StringUtils.isBlank(notificationResponseBean.getKey())) {
                return Response.status(Response.Status.CREATED).build();
            }
            return Response.status(Response.Status.CREATED).entity(notificationResponseBean.getKey()).build();
```

Reply:
Fixed with 6cb9adbac9ab6eab76f2f39ad85fb5151201161f